### PR TITLE
DBZ-4295: Support stop_on_reshard flag

### PR DESCRIPTION
**Pull request by sonne5**
Closes https://issues.redhat.com/browse/DBZ-4295

### CODE REVIEW

**Comment by Naros** [comment]
Hi @sonne5 could you please re-word the commit message to begin with `DBZ-4295`?

**Comment by sonne5** [comment]
> Hi @sonne5 could you please re-word the commit message to begin with `DBZ-4295`?
sure thing, thanks @Naros

**Review by keweishang — APPROVED** [review]
LGTM 👍

**Comment by keweishang** [comment]
From the ticket, @sonne5 mentioned:
> If this is `true`, say we are reshard A into B and C, the stream on shard A will be stopped.

If I understood correctly, if this is set to `true`, after resharding, the stream stops on shard A and does not automatically start on shard B and shard A. Essentially, the connector stops streaming from the entire keyspace. How does the connector resume streaming from the given keyspace (e.g. restarting the connector)?

Thanks!

**Comment by sonne5** [comment]
@keweishang thanks for reviewing!

> How does the connector resume streaming from the given keyspace (e.g. restarting the connector)?

So assuming a keyspace with one shard and we have one connector for it. This shard split into 2 shards, with the flag set to true, we need to issue post requests to connect to:
1. stop the connector for the old shard (assuming the old connector won't stop by itself?)
2. start new connectors for the new shards (via different config properties `vitess.shard`)

Does it make sense?

**Comment by keweishang** [comment]
>stop the connector for the old shard (assuming the old connector won't stop by itself?)
Yes. I think the connector won't automatically stop, it simply won't receive new data since the `stop_on_reshard` flag is set to `true`. So you'll need to send a POST request to stop or to remove the old connector.

>start new connectors for the new shards (via different config properties vitess.shard)
You need to be a bit more careful here. You can create two new connectors, one for shard B and one for shard C. Each new connector will start streaming from their corresponding shard. However since they're new connectors, no VGTID offset is stored in the Kafka topic, therefore, they'll start from the latest VGTID for shard B and C. If Shard B and C have any transactions between the moment of the resharding and the moment when the new connectors are created, those transactions won't be streamed to the new connectors. To put it in another way, the new connectors don't have VGTID offset saved and therefore can only receive transactions that happened after the connectors' creation.

**Comment by sonne5** [comment]
> no VGTID offset is stored in the Kafka topic, therefore, they'll start from the latest VGTID for shard B and C. 

This is something I would like get your thoughts as well. I think it is reasonable to ask the reshard workflow to record the cutover GTID somewhere so that in theory we can get to know what's the position that we should start for the new shards after the reshard. 

Looking at the code, the `restartVgtid` is stored inside [SourceInfo](https://github.com/debezium/debezium-connector-vitess/blob/ef0cc917d8348114f60261496f98bac4b721e3cf/src/main/java/io/debezium/connector/vitess/SourceInfo.java#L30) - I'm thinking add another `vitess.gtid` in the config so that if it is not empty we init the vstream request with it (instead of using `CURRENT`: https://github.com/debezium/debezium-connector-vitess/blob/c0d12be3cfe7db29d6cbb1b4c1b258153e681632/src/main/java/io/debezium/connector/vitess/connection/VitessReplicationConnection.java#L216

Does it sound reasonable to you? If so, I can open another PR for that small change.

**Comment by sonne5** [comment]
(^ I cut https://issues.redhat.com/browse/DBZ-4297 for the default starting point)

**Comment by sonne5** [comment]
@keweishang I thought a bit more, it does not make a lot of sense to allow setting `stop_on_resharding = true` while we only allow starting from `current` (latest) position for a shard (the exact reason you called out earlier).

I added another config `vitess.gtid` so that when the shard is set, we allow the starting gtid in the config. If it is not set, we fall back to the vtctld approach.

Let me know if this makes sense to you, happy to discuss more if needed.

**Comment by keweishang** [comment]
>Does it sound reasonable to you? If so, I can open another PR for that small change.

Yes. It sounds reasonable. If I understand correctly, this new configuration `vitess.gtid` will be used only when creating the new connector, after the new connector has been created and running for a while, new transaction's VGITD offset will be saved in Kafka topic and the configuration `vitess.gtid` will not be useful anymore. Sounds good to me. I recommend documenting the new configuration `vitess.gtid` behaviour [here](https://debezium.io/documentation/reference/connectors/vitess.html#vitess-connector-properties), together with the `stop_on_reshard` configuration.

**Comment by keweishang** [comment]
The source code for the Vitess Connector Documentation can be found here: https://github.com/debezium/debezium/blob/main/documentation/modules/ROOT/pages/connectors/vitess.adoc

**Comment by sonne5** [comment]
> this new configuration vitess.gtid will be used only when creating the new connector, after the new connector has been created and running for a while, new transaction's VGITD offset will be saved in Kafka topic and the configuration vitess.gtid will not be useful anymore. 

Yep, you are correct. Happy to update the document after this is merged.

Do you happen to have the permission to submit this PR and do you happen to know what's the release process for this project?

**Comment by keweishang** [comment]
>Do you happen to have the permission to submit this PR and do you happen to know what's the release process for this project?

I haven't been involved in recent releases. @Naros this PR looks good to me. Can you have a final review and merge the PR?

**Comment by sonne5** [comment]
@keweishang unrelated to this PR, just a noob question on vitess connector: how does it handle the restart? Ignore this change, the current behavior is we call vtctld to get current gtid positoin and start the vstream - which seems it is not checking the offset topic.

Intuitively, I would think the thing you described to happen: we only call vtctld to get GTID for the first time, after that we will just consume the offset topic. But looks like we always call VStream endpoint without checking that topic - what do I miss? If you can share some related code pointers, that would be awesome!

File: pom.xml
Comment (gunnarmorling) [inline pom.xml:25]: Could you instead please update the version in the [Debezium BOM](https://github.com/debezium/debezium/blob/main/debezium-bom/pom.xml#L26-L31)? All the dependency versions are coordinated via that one.

File: pom.xml
Comment (gunnarmorling) [inline pom.xml:150]: Should go away, see above.

File: src/main/java/io/debezium/connector/vitess/VitessConnectorConfig.java
Comment (gunnarmorling) [inline src/main/java/io/debezium/connector/vitess/VitessConnectorConfig.java:209]:
```suggestion
    public String getGtid() {
```

Unless it's a static member, we use that style in general for abbreviations in member names (it's not done consistently 100% unfortunately).

File: src/main/java/io/debezium/connector/vitess/VitessConnectorConfig.java
Comment (gunnarmorling) [inline src/main/java/io/debezium/connector/vitess/VitessConnectorConfig.java:85]: I think we can omit the default and just react to the value if it's not `null`?

File: src/main/java/io/debezium/connector/vitess/connection/VitessReplicationConnection.java
Comment (gunnarmorling) [inline src/main/java/io/debezium/connector/vitess/connection/VitessReplicationConnection.java:243]:
```suggestion
            LOGGER.info("VGTID '{}' is set to the GTID {} for keyspace: {} shard: {}", gtid, startingGtid, config.getKeyspace(), shard);
```

**Review by gunnarmorling — COMMENTED** [review]
Thanks for the change, @sonne5, and thanks for reviewing, @keweishang! Some comments inline. When you do the suggested PR for the BOM change in the main repo, you could also add the docs addition as part of that one. In the context of that main repo PR, this PR's branch will also be built (assuming they are done using the same branch name in both repos).

**Comment by gunnarmorling** [comment]
> What's the release process for this project?

Once those fixes suggested above are done, we can merge the PRs (this repo, and main one), and they will be part of the 1.8.0.Beta1 release, which is scheduled for Nov. 25th.

We're looking for more maintainers for the Vitess connector btw, so if you got more changes in mind, your contributions will be more than welcomed. Thanks a lot!

File: pom.xml
Comment (sonne5) [reply pom.xml:25]: Thanks for the pointer! Since the flag is not available in Vitess 9.0.0 release, we need this PR first https://github.com/debezium/debezium/pull/2951

File: src/main/java/io/debezium/connector/vitess/VitessConnectorConfig.java
Comment (sonne5) [reply src/main/java/io/debezium/connector/vitess/VitessConnectorConfig.java:85]: yep, done

File: pom.xml
Comment (gunnarmorling) [reply pom.xml:25]: Merged that PR. Note if both PRs would have used the same branch name, this PR here would have been built against that update over there, without having to merge that other PR first. That's a nice way to avoid merging the upstream PR first and then noticing it caused issues and needs backing out. In this case it was not a problem, as we had seen that the update to 12 works here already. So just mentioning it as some context info :)

Comment (sonne5) [reply pom.xml:25]: Ah, didn't know that. TIL

Do you know how do I bump the reference in this branch to pick up the version change? I removed this line and tests are failing.

Comment (sonne5) [reply pom.xml:25]: Or is it OK if I remove this two lines in pom.xml and we merge as is?

Comment (gunnarmorling) [reply pom.xml]: I hope it will pass actually. This build fetches the main branch from the core repo (with your version update) and builds against that (in absence of a branch with the same name over there).

**Comment by gunnarmorling** [comment]
> I do have another idea that we can improve the connector, see some discussion here

Makes sense, PR welcome :)

> looking forward to interacting more on this project!

Same!

Would also love a PR for getting your org added to the [list of Debezium users](https://debezium.io/community/users/), should you be interested 🙏 .

File: pom.xml
Comment (sonne5) [reply pom.xml]: Ah yeah, you are right! I guess it's just my local branch issue

**Comment by gunnarmorling** [comment]
Rebased and applied; thx, @sonne5!

**Pull request merged**
